### PR TITLE
Includes simulation tips section in the docs

### PR DESCRIPTION
**Pull request by tomchor**
Closes https://github.com/CliMA/Oceananigans.jl/issues/1478

As mentioned in https://github.com/CliMA/Oceananigans.jl/issues/1542 I couldn't build the docs locally so I probably made some markdown mistakes along the way, which is why this is a draft pull request. This first draft is also somewhat incomplete.

I figured I'd create the PR early to get some feedback along the way though. I'm especially unfamiliar with the part about converting arrays to CUDA etc., so some help/collaboration there is much appreciated.

CC @ali-ramadhan @glwagner @navidcy @francispoulin

### CODE REVIEW

File: docs/src/simulation_tips.md
Comment (francispoulin) [inline docs/src/simulation_tips.md:114]: You did a great job in explaining @inline above.  Is it worth having a sentence explaining @inbounds as well?

File: docs/src/simulation_tips.md
Comment (francispoulin) [inline docs/src/simulation_tips.md:119]: Should the comments be changed to describe the functio to be interpolated?

File: docs/src/simulation_tips.md
Comment (francispoulin) [inline docs/src/simulation_tips.md:106]: Very nice example by the and this has helped me to understand this better

File: docs/src/simulation_tips.md
Comment (francispoulin) [inline docs/src/simulation_tips.md:141]: There is an nvidia command that helps to profile.  Should we mention an example so the user has an idea where to start?

File: docs/src/simulation_tips.md
Comment (francispoulin) [inline docs/src/simulation_tips.md:143]: instead of resolution do you mean accuracy?

**Review by francispoulin — COMMENTED** [review]
I think this is a reallly nice contribution and I learned some things along the way so thanks.

I have still a lot to learn about efficiency so will let someone else who is more knowledgeable on the topic approve and add more feedback

File: docs/src/simulation_tips.md
Comment (glwagner) [inline docs/src/simulation_tips.md:51]: It may be worth mentioning that `KernelAbstractions` "force inlines" all GPU code. I can't remember if CPU code is also force inlined. @vchuravy and @jakebolewski have give advice regarding the use of `@inline` in the past. Force-inlining suggests that we don't always need the annotation `@inline` but for some reason it still pervades all our code... ?

File: docs/src/simulation_tips.md
Comment (glwagner) [inline docs/src/simulation_tips.md:145]:
```suggestion
Use the [`nvidia-smi`)[https://developer.nvidia.com/nvidia-system-management-interface] command line utility to monitor the memory usage of the GPU.
```

**Comment by glwagner** [comment]
Thanks for starting this @tomchor .

**Comment by ali-ramadhan** [comment]
Docs should show up at https://clima.github.io/OceananigansDocumentation/previews/PR1543/ once a commit has been made after the PR has been opened.

File: docs/src/simulation_tips.md
Comment (ali-ramadhan) [reply docs/src/simulation_tips.md:51]: I added a suggestion to mention KernelAbstractions.jl but it might not be useful to discuss inlining behavior to new users (many of whom are completely new to Julia).

My (probably unpopular) opinion is that I like to explicitly `@inline` functions to tell people reading my code that this is a performance-critical function I intend to have inlined whenever possible (even if it'll be force-inlined anyways). But this is probably a debate for another place.

File: docs/src/simulation_tips.md
Comment (ali-ramadhan) [inline docs/src/simulation_tips.md:136]: Is it worth also mentioning https://github.com/CliMA/Oceananigans.jl/issues/1241 ?

It's possible that the issue of big abstract operations on the GPU will be resolved in the future.

File: docs/src/simulation_tips.md
Comment (ali-ramadhan) [inline docs/src/simulation_tips.md:152]: Should we also talk about CUDA scalar operations (https://juliagpu.github.io/CUDA.jl/dev/usage/workflow/#UsageWorkflowScalar) and how/when to use `CUDA.allowscalar` and `CUDA.@allowscalar`?

File: docs/src/simulation_tips.md
Comment (tomchor) [reply docs/src/simulation_tips.md:114]: I don't know much about `@inbounds` in Julia, but from what I can tell the user doesn't need to use it that much (except when creating `KernelComputedField`s, which has many other difficulties). So I'd say we don't need to touch on that here. But I might be missing something.

File: docs/src/simulation_tips.md
Comment (tomchor) [reply docs/src/simulation_tips.md:106]: Thanks!

File: docs/src/simulation_tips.md
Comment (tomchor) [reply docs/src/simulation_tips.md]: Done!

File: docs/src/simulation_tips.md
Comment (tomchor) [reply docs/src/simulation_tips.md:52]: I think the docs on inlining in Julia are a bit scarce. There's a lot of spread out in discourse comments. For the time being I think this is the limit of my knowledge. But feel free to adapt the text if you think it needs more details.

File: docs/src/simulation_tips.md
Comment (tomchor) [reply docs/src/simulation_tips.md:153]: My guess is yes, but I know nothing about this topic, so some collaboration would be very much appreciated :)

File: docs/src/simulation_tips.md
Comment (tomchor) [reply docs/src/simulation_tips.md:137]: > It's possible that the issue of big abstract operations on the GPU will be resolved in the future.

Hopefully! Do you have any insider information about that? I haven't seen much discussed in `KernelAbstractions.jl`

**Comment by tomchor** [comment]
> Docs should show up at https://clima.github.io/OceananigansDocumentation/previews/PR1543/ once a commit has been made after the PR has been opened.

This gives me a 404 error!

File: docs/src/simulation_tips.md
Comment (tomchor) [reply docs/src/simulation_tips.md:146]: Yes! Thanks

**Comment by navidcy** [comment]
> > Docs should show up at https://clima.github.io/OceananigansDocumentation/previews/PR1543/ once a commit has been made after the PR has been opened.
> 
> This gives me a 404 error!

I believe you have to open the PR (not have it as "Draft") and then it'll push the preview....

File: docs/src/simulation_tips.md
Comment (glwagner) [reply docs/src/simulation_tips.md:52]: I think its fine to recommend `@inline` for style / clarity. But we shouldn't mislead people in the docs by saying that `@inline` is necessary for performance if it has no effect.

Comment (tomchor) [reply docs/src/simulation_tips.md:52]: @glwagner are you suggesting that `@inline` doesn't impact performance?

**Comment by tomchor** [comment]
I finished the first of the things I know about. I haven't writen the very last subsection though, which is about viewing/using arrays in GPU runs because honestly I don't know enough to write about it. I know there's a function called `view()` that helps with slicing, and you can also use `adapt` to view CUDA Arrays in REPL for investigating, but I feel like I don't know enough of this aspect of the code to write this. There are two options here:

- Someone helps me out and writes that last section
- Or I can merge the PR without it for now (there are many other important things there for the user) and we worry about that last part in a future PR

Thoughts?

File: docs/src/simulation_tips.md
Comment (francispoulin) [reply docs/src/simulation_tips.md:135]: Sadly, I don't remember when to use @inbuonds and when not to use it.  @ali-ramadhan explained it to me once.  Maybe he could help in adding a sentence?  I would certainly use this when writing code as at the moment it seems mysterious to me when you use it and when you don't.

Comment (navidcy) [reply docs/src/simulation_tips.md:135]: It just tell julia to not bother checking  whether the index is beyond the array’s limit. At least that’s what I think it does...

https://docs.julialang.org/en/v1/devdocs/boundscheck/

**Comment by glwagner** [comment]
I'm wondering if we should provide a separate page on "Using GPUs"? While the simulation tips for CPUs are really performance optimizations that are optional, the GPU simulation tips are mostly required to run without errors. There's a few other things that are required to get things working on GPUs --- for example, Oceananigans must be _built_ (not just run) with a GPU / CUDA installation available; this is a common pitfall on clusters.

**Comment by tomchor** [comment]
> I'm wondering if we should provide a separate page on "Using GPUs"? While the simulation tips for CPUs are really performance optimizations that are optional, the GPU simulation tips are mostly required to run without errors.

That's a good point. Although I think we could avoid creating another page and put that information in the ["Using GPUs"](https://clima.github.io/OceananigansDocumentation/stable/using_gpus/) page, so that things are more condensed.

**Comment by glwagner** [comment]
> > I'm wondering if we should provide a separate page on "Using GPUs"? While the simulation tips for CPUs are really performance optimizations that are optional, the GPU simulation tips are mostly required to run without errors.
> 
> That's a good point. Although I think we could avoid creating another page and put that information in the ["Using GPUs"](https://clima.github.io/OceananigansDocumentation/stable/using_gpus/) page, so that things are more condensed.

🤦 there's already Using GPUs of course...

**Comment by navidcy** [comment]
We have a preview!
https://clima.github.io/OceananigansDocumentation/previews/PR1543/simulation_tips/
Seems like the section needs to be finished (see, e.g., last sentence...). @tomchor?

File: docs/src/simulation_tips.md
Comment (navidcy) [inline docs/src/simulation_tips.md:153]: I'm a bit hesitant to reference the users to a package that is neither registered nor tested. (mostly the latter tbh)
What do others think?

**Review by navidcy — CHANGES_REQUESTED** [review]

File: docs/src/simulation_tips.md
Comment (ali-ramadhan) [reply docs/src/simulation_tips.md:153]: Oceanostics.jl is registered. It could do with some actual tests but plenty of Oceananigans.jl functionality is not tested.

I think it's good to link to packages outside of Oceananigans.jl to give readers the idea that creating their own packages for extra functionality is okay (and actually encouraged!).

I would even advocate for creating an extra page in the docs to link to other packages related to Oceananigans.jl and being liberal with the packages we include.

**Comment by tomchor** [comment]
Thanks for the help with this PR, @navidcy! Much appreciated

> Seems like the section needs to be finished (see, e.g., last sentence...). @tomchor?

Ah, that's the limit of my knowledge on GPUs. We could either drop that session for now or [someone else could write it](https://github.com/CliMA/Oceananigans.jl/pull/1543#issuecomment-814533032). I really don't think I know enough to write that section.

File: docs/src/simulation_tips.md
Comment (ali-ramadhan) [inline docs/src/simulation_tips.md:193]:
```suggestion
On the CPU Oceananigans.jl uses regular `Array`s, but on the GPU it has to use `CuArray`s
from the CUDA.jl package. You might need to keep this difference in mind when using arrays
to `set!` initial conditions or when using arrays to provide boundary conditions and
forcing functions.

To learn more about working with `CuArray`s, see the
[array programming](https://juliagpu.github.io/CUDA.jl/dev/usage/array/) section
of the CUDA.jl documentation.

Something to keep in mind when working with `CuArray`s is that you do not want to set or
get/access elements of a `CuArray` outside of a kernel. Doing so invokes scalar operations
in which individual elements are copied from or to the GPU for processing. This is very
slow and can result in huge slowdowns. For this reason, Oceananigans.jl disables CUDA
scalar operations by default.

See the [scalar indexing](https://juliagpu.github.io/CUDA.jl/dev/usage/workflow/#UsageWorkflowScalar)
section of the CUDA.jl documentation for more information on scalar indexing.

Sometimes you need to perform scalar operations on `CuArray`s in which case you may want
to temporarily allow scalar operations with the `CUDA.@allowscalar` macro or by calling
`CUDA.allowscalar(true)`.
```

**Review by ali-ramadhan — APPROVED** [review]
I added some quick text under the last section. Should be better than nothing.

Happy to merge this as-is now since it's already a great addition to the docs that we can continue expanding in the future as common GPU issues pop up.

File: docs/src/simulation_tips.md
Comment (navidcy) [reply docs/src/simulation_tips.md:193]: Great.

**Review by navidcy — APPROVED** [review]

**Comment by ali-ramadhan** [comment]
I think GPU tests crapped out but docs are still building so as long as the PR docs preview looks good, then this should be good to merge!

**Comment by tomchor** [comment]
@ali-ramadhan currently waiting for the preview to load so that I can merge, but I don't think it'll do that with the GPU test failing. How do we fix that?

**Comment by ali-ramadhan** [comment]
@tomchor Ah are you okay with committing my suggestion before merging? 

I think docs did get built and deployed:
https://buildkite.com/clima/oceananigans/builds/2048#0d49a5b8-1dec-4f7b-b82a-cf6f8809e234/40-574
https://clima.github.io/OceananigansDocumentation/previews/PR1543/simulation_tips/

Working on getting you guys access to Buildkite so you can control it as well. I'm not a Buildkite admin so I have to ask one to invite other people...

**Comment by tomchor** [comment]
> @tomchor Ah are you okay with committing my suggestion before merging?

Sorry, I forgot to do that!

**Comment by ali-ramadhan** [comment]
Everything passed 🎉 Thanks so much for adding this great information to the docs @tomchor!

File: docs/src/simulation_tips.md
Comment (glwagner) [reply docs/src/simulation_tips.md]: I don't think it does for GPU, but it might for CPU...

Comment (glwagner) [reply docs/src/simulation_tips.md]: Could be fun to illustrate these points with code snippets that do some benchmarking that users can use for themselves (some day).****

**Comment by tomchor** [comment]
Thanks, everyone. I'll merge this for now and we can improve on it later based on feedback from users.

@glwagner I'm thinking of opening another PR soon to address your [comment about using GPUs](https://github.com/CliMA/Oceananigans.jl/pull/1543#issuecomment-818076711). Like I mentioned, some of this info is already available in the "Using GPUs" page, but maybe it's useful to expand a bit on it and link this newly created page there.

**Pull request merged**
